src/Analyser/MutatingScope.php

```diff
@@ -1111,6 +1111,39 @@ private function resolveType(Expr $node): Type
 		}
 
 		if ($node instanceof Node\Expr\BinaryOp\Mod || $node instanceof Expr\AssignOp\Mod) {
+			if ($node instanceof Node\Expr\AssignOp) {
+				$left = $node->var;
+				$right = $node->expr;
+			} else {
+				$left = $node->left;
+				$right = $node->right;
+			}
+
+			$leftType = $this->getType($left);
+			$rightType = $this->getType($right);
+
+			$integer = new IntegerType();
+			$positiveInt = IntegerRangeType::fromInterval(0, null);
+			if ($integer->isSuperTypeOf($rightType)->yes()) {
+				$rangeMin = null;
+				$rangeMax = null;
+				if ($rightType instanceof IntegerRangeType) {
+					$rangeMax = $rightType->getMax() !== null ? $rightType->getMax() - 1 : null;
+				} elseif ($rightType instanceof ConstantIntegerType) {
```

Comment:
I think this will lead to creating `int<X, max>` when you have `1|2|3` which is wrong I guess.

Reply:
another great catch. added a test and a fix

Reply:
That's because `instanceof *Type` is almost always wrong and you can find edge-cases like that.. https://phpstan.org/developing-extensions/type-system#querying-a-specific-type

Reply:
Right now it's still wrong for unions of disjoint ranges like `int<min, 3>|int<4, max>`.

Reply:
got this case and a few more also covered. I guess you won't like that, I had to add a few more `instanceof *Type` calls.

I somehow get the feeling you let me built this thing intentionally and we will now refactor the logic regarding merging `IntegerRangeType` and `ConstantIntegerType` types (or unions of it) into either `TypeUtils` or `IntegerRange` class, right? ;)

Reply:
I guess it's fine like that. A correct approach is to add use-case based methods on `PHPStan\Type\Type` that would tell us exactly what we need here. Each `Type` implementation would return what it knows.

I'm preparing myself mentally for this refactoring for a long time, in the end `PHPStan\Type\Type` might have hundreds of methods, and it might be the only valid situation considered quality code in the universe 😂